### PR TITLE
New Typing Files for forever-monitor

### PR DESCRIPTION
**Pull request by shuntksh**
Please fill in this template.

- [X] Make your PR against the `master` branch.
- [X] Use a meaningful title for the pull request. Include the name of the package modified.
- [X] Test the change in your own code. (Compile and run.)
- [X] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped#make-a-pull-request).
- [X] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped#common-mistakes).
- [X] Run `tsc` without errors.
- [X] Run `npm run lint package-name` if a `tslint.json` is present.

Select one of these and delete the others:

If adding a new definition:
- [X] The package does not provide its own types, and you can not add them.
- [X] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [X] Create it with `npm run new-package package-name`, not by basing it on an existing project.
- [X] `tslint.json` should be present, and `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, and `strictNullChecks` set to `true`.

### CODE REVIEW

File: forever-monitor/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+
+interface forever {
+    Monitor: Monitor;
+    start: (script: string, options: Options) => Monitor;
+    kill: (pid: number, killTree?: boolean, signal?: string, callback?: () => any) => void;
+    checkProcess: (pid: number) => boolean;
+    version: string;
+}
+
+export default forever;
```
Comment (DanielRosenwasser) [inline forever-monitor/index.d.ts]: This only creates a type that is importable as a default, not a value.

File: forever-monitor/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,104 @@
+// Type definitions for forever-monitor 1.7
+// Project: https://github.com/nodejitsu/forever-monitor#readme
+// Definitions by: Shun Takahashi <https://github.com/shuntksh>
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+/// <reference types="node" />
+
+interface spawnWith {
```
Comment (DanielRosenwasser) [inline forever-monitor/index.d.ts]: Capitalize type names please.

**Review by DanielRosenwasser — COMMENTED** [review]
The problem with how this is structured is that the other entities are not exported at all, however, you want to use the `export =` syntax because the original module is a CommonJS module.

I would suggest revising this as follows:

```
declare namespace forever {
    export class Monitor {
        // ...
    }

    export interface SpawnWith {
        // ...
    }

    export const version: string
}

export = forever;
```

File: forever-monitor/index.d.ts [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+    customFds: number[];
+    setsid: boolean;
+    uid: number;
+    gid: number;
+}
+
+interface parserArgs {
```
Comment (DanielRosenwasser) [inline forever-monitor/index.d.ts]: Where is this used?

File: forever-monitor/index.d.ts [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+
+    /**
+     * @param command - Command string to parse
+     * @param args - Additional default arguments
+     */
+    parseCommand(command: string, args?: string[]): (false | { command: string, args?: string[]});
+}
+
+interface forever {
+    Monitor: Monitor;
```
Comment (DanielRosenwasser) [inline forever-monitor/index.d.ts]: Is this a `Monitor`? Or is its type `typeof Monitor`?

**Review by DanielRosenwasser — CHANGES_REQUESTED** [review]

File: forever-monitor/index.d.ts (hunk 4 above)
Comment (shuntksh) [inline forever-monitor/index.d.ts]: This would be class itself not typeof class. So new forever.Monitor(args...); is the actual usage.

**Comment by shuntksh** [comment]
@DanielRosenwasser Thank you for your great suggestion. I quickly tried to reorganize the typing with namespace model as you suggested. Updated typings also worked as expected, however, now test is failing due to "export-just-namespace" warning:

```bash
Error in forever-monitor
Command failed: node /Workspace/_forked/DefinitelyTyped/node_modules/tslint/lib/index.js/../tslint-cli.js --format stylish index.d.ts

index.d.ts
96:1  export-just-namespace  Instead of `export =`-ing a namespace, use the body of the namespace as the module body.
```

Do I have to declare another object / interface inside the namespace and export it instead like below?

```ts
declare namespace Forever {
    interface forever {
       // some typing
    }
}
declare const forever: Forever.forever;
export = forever;
```

**Comment by shuntksh** [comment]
@DanielRosenwasser I also tried flat-export model to avoid lint-ing error around exporting namespace. And fixed points you mentioned in review with adding few more test cases to test file. Can you take a look?

**Review by DanielRosenwasser — APPROVED** [review]

**Comment by DanielRosenwasser** [comment]
Sorry about that advice, our lint rules aren't big on it. Looks good now though, thanks!

**Comment by janwo** [comment]
@shuntksh It might be great having types of the forever module to call `startServer` properly. 👍

**Pull request merged**